Megaparsec/Megaparsec.lean

```diff
@@ -288,8 +315,178 @@ def msₜ [m: Monad M]: Monad (StateT σ M) :=
+        let reerr _ _ := cerr err ms
+        let p₁ := r err
+        p₁.unParser B ms rcok rcerr reok reerr
+    let meerr err ms :=
+        let rcok x s' _ := cok x s' (toHints s'.stateOffset err)
+        let rcerr _ _ := eerr err ms
+        let reok x s' _ := eok x s' (toHints s'.stateOffset err)
+        let reerr _ _ := eerr err ms
+        let p₁ := r err
+        p₁.unParser B ms rcok rcerr reok reerr
+    p.unParser B s cok mcerr eok meerr
+
+def pEof [stream : Stream S] : @ParsecT S M E stream m Unit :=
+  ParsecT.mk $ fun _ s _ _ eok eerr =>
+    let input := s.stateInput
+    let o := s.stateOffset
+    let pst := s.statePosState
+    let de := s.stateParseErrors
+    match stream.take1 input with
+      | Option.none => eok Unit.unit s []
+      | Option.some (x,_) =>
+          let us := (Option.some ∘ ErrorItem.tokens ∘ nes) x
+          let ps := [ ErrorItem.eof ]
+          eerr (ParseError.trivial o us ps) (State.mk input o pst de)
+
+instance (E S : Type) [m : Monad M] [stream : Stream S]
+         [o : Ord stream.Token] [e : BEq stream.Token] :
+         @MonadParsec (@ParsecT S M E stream m) E S (@mprsₜ S M E stream m) (@altpₜ S M E stream o e m) stream where
+  parseError _ err := ParsecT.mk $ fun _ s _ _ _ eerr => eerr err s
+  label := pLabel
+  attempt _ p :=
+    ParsecT.mk $ fun B s cok _ eok eerr =>
+      let eerr' err _ := eerr err s
+      p.unParser B s cok eerr' eok eerr'
+  lookAhead _ p :=
+    ParsecT.mk $ fun B s _ cerr eok eerr =>
+      let eok' a _ _ := eok a s []
+      p.unParser B s eok' cerr eok' eerr
+  notFollowedBy := pNotFollowedBy
+  withRecovery := pWithRecovery
+  observing _ p := ParsecT.mk $ fun B s cok _ eok _ =>
+    let cerr' err s' := cok (Either.left err) s' []
+    let eerr' err s' := eok (Either.left err) s' (toHints s'.stateOffset err)
+    p.unParser B s (cok ∘ Either.right) cerr' (eok ∘ Either.right) eerr'
+  eof := pEof
+  token A matcher ps := ParsecT.mk $ fun B s cok _ _ eerr =>
+    let input := s.stateInput
+    let o := s.stateOffset
+    let pst := s.statePosState
+    let de := s.stateParseErrors
+    match stream.take1 input with
+      | .none => eerr (.trivial o (.some .eof) ps) s
+      | .some (c, cs) => match matcher c with
+        | .none =>
+            let us := (.some ∘ .tokens ∘ nes) c
+            eerr (.trivial o us ps) s
+        | .some x => cok x (State.mk cs (o + 1) pst de) []
+  tokens A matcher chunk := ParsecT.mk $ fun B s₀ cok _ eok eerr =>
+    let unexpect pos' u :=
+      let us := pure u
+      -- let ps := [ (@ErrorItem.tokens stream.Token chunk) ]
+      let es := match stream.chunkToTokens chunk with
+        | List.cons x xs => [ErrorItem.tokens $ NonEmptyList.cons x xs]
+        | [] => [ErrorItem.label $ NonEmptyList.cons 'e' ['m', 'p', 't', 'y', ' ', 't', 'a', 'r', 'g', 'e', 't', ' ', 'c', 'h', 'u', 'n', 'k']]
+        -- ^ This should never happen, because an empty target always succeeds by design
```

Comment:
```suggestion
        | [] => unreachable!
```
If this should never happen, then you can make it panic with `unreachable!` so you know something went wrong with your assumptions

Reply:
I kind of agree, but also, I strongly disagree. In languages without fault isolation, we should be very careful when we crash.

Here I don't care about this assumption much. If for some reason, the target chunk is empty and the actual body of the function didn't handle it, I'm ok with erroring out [because the semantics of empty chunks in `tokens` are arbitrary](https://yatima.zulipchat.com/#narrow/stream/326780-meta/topic/Parsec.20pair.20prog/near/286805600).

If the semantics would have been unambiguous, it would be fair to make a partial function here. That's why I mention "by design", BTW!

Finally, with partial functions (and panicking in a branch is partial), I see only one truly good design (again, in languages without fault isolation): an empty `Partial` typeclass and forcing the users to explicitly do unsafe extraction from `Partial a` to `a`, or stay in the `Partial` world (see PureScript).